pkg/heartbeat/heartbeat.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package heartbeat
```

Comment:
I assume the podmaster is trying to do master election for the apiserver.

Right now, the heartbeat here is only added to the controller-manager and scheduler, but I imagine something like it will also need to be on the apiserver itself, especially if there are multiple instances. Hypothetically the apiserver could use an etcd-backed client, rather than a rest-based one. But right now the impl here is very simple, and doesn't do any sort of leader election. 

Hypothetically the Components api could be modified to add a Leader field, since etcd handles rudimentary synchronization. But afaik, etcd has no transactions. So I don't think you could read the component list and write in the same atomic action. It looks like the podmaster works because the leader is read/write from the same etcd field.